src/avm2/natives/xml.ts

```diff
@@ -347,82 +353,83 @@ module Shumway.AVM2.AS {
       throw new TypeError(formatErrorMessage(Errors.ConvertUndefinedToObjectError));
-    } else if (v instanceof ASXML) {
+    }
+    if (v instanceof ASXML) {
       return v;
-    } else if (v instanceof ASXMLList) {
-      if (v.length() === 1) {
-        return v._children[0];
-      }
-      throw new TypeError(formatErrorMessage(Errors.XMLMarkupMustBeWellFormed));
-    } else {
-      var x = xmlParser.parseFromString(String(v));
-      if (x.length() === 0) {
-        var x = new XML(ASXMLKind.Text);
-        return x;
-      } else if (x.length() === 1) {
-        x._children[0]._parent = null;
-        return x._children[0];
+    }
+    if (v instanceof ASXMLList) {
+      if (v._children.length !== 1) {
+        Runtime.throwError('TypeError', Errors.XMLMarkupMustBeWellFormed);
       }
-      throw "SyntaxError in ToXML";
+      return v._children[0];
+    }
+    // The E4X spec says we must throw a TypeError for non-Boolean, Number, or String objects.
+    // Flash thinks otherwise.
+    var x = xmlParser.parseFromString(asCoerceString(v));
+    var length = x.length();
+    if (length === 0) {
+      return createXML(ASXMLKind.Text);
     }
+    if (length === 1) {
+      x._children[0]._parent = undefined;
+      return x._children[0];
+    }
+    Runtime.throwError('TypeError', Errors.XMLMarkupMustBeWellFormed);
   }
 
   // 10.4 ToXMLList
-  function toXMLList(value) {
+  function toXMLList(value: any, targetList: ASXMLList): void {
+    // toXMLList is supposed to just return value if it's an XMLList already. For optimization
+    // purposes, we handle that case at the callsites.
+    release || assert(!(value instanceof ASXMLList));
     if (value === null) {
-      throw new TypeError(formatErrorMessage(Errors.ConvertNullToObjectError));
-    } else if (value === undefined) {
-      throw new TypeError(formatErrorMessage(Errors.ConvertUndefinedToObjectError));
-    } else if (value instanceof XML) {
-      var xl = new XMLList(value.parent, value.name);
-      xl.appendChild(value);
-      return xl;
-    } else if (value instanceof XMLList) {
-      return value;
-    } else {
-      var parentString = '<parent xmlns=\'' + ASXML.defaultNamespace + '\'>' +
-        value + '</parent>';
-      var x = toXML(parentString);
-      var xl = new XMLList();
-      for (var i = 0; i < x.length(); i++) {
-        var v = (<any> x)._children[i];
-        v._parent = null;
-        xl.appendChild(v);
-      }
-      return xl;
+      Runtime.throwError('TypeError', Errors.ConvertNullToObjectError);
+    }
+    if (value === undefined) {
+      Runtime.throwError('TypeError', Errors.ConvertUndefinedToObjectError);
+    }
+    if (value instanceof ASXML) {
+      targetList.append(value);
+      return;
+    }
+    // The E4X spec says we must throw a TypeError for non-Boolean, Number, or String objects.
+    // Flash thinks otherwise.
+    var parentString = '<parent xmlns="' + ASXML.defaultNamespace + '">' + value + '</parent>';
```

Comment:
it might was wrong before, but we shall escapeAttributeValue for `ASXML.defaultNamespace`